jupyter_client/threaded.py

```diff
@@ -80,6 +82,29 @@ def stop(self) -> None:
 
     def close(self) -> None:
         """ "Close the channel."""
+        if self.stream is not None and self.ioloop is not None:
+            # c.f.Future for threadsafe results
+            f = Future()
+
+            def close_stream():
+                try:
+                    self.stream.close()
+                    self.socket.close(linger=0)
+                    self.socket = None
+                except Exception as e:
+                    f.set_exception(e)
+                else:
+                    f.set_result(None)
+
+            self.ioloop.add_callback(close_stream)
+            try:
+                f.result(timeout=5)
```

Comment:
This pattern should prevent hangs in case of errors where we use `Event.wait()` without timeouts elsewhere in this file.